### PR TITLE
Polars: auto-calibrate noise scale in bounded-DP

### LINKED ISSUE #1941: Polars: bounded-DP auto-calibration of scale parameter
`.dp.mean` fails to auto-calibrate unless you pass a bogus scale. This is because d_in is configured to 1, but the smallest valid dataset distance is 2 in bounded-DP.

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1941
This makes it possible to run the Docs notebook at the tip of the stack.

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1943?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1943** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1943?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1942** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1942?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1920** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1920?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/measurements/make_private_expr/expr_noise/mod.rs [hunk 1]
```diff
@@ -182,7 +183,11 @@ where
         None => {
             // when scale is unknown, set it relative to the sensitivity of the query
             let margin = input_domain.active_margin().cloned().unwrap_or_default();
-            t_prior.map(&(margin.l_0(1), 1, margin.l_inf(1)))?
+            let d_in = match margin.public_info {
+                Some(MarginPub::Lengths) => 2,
```
Comment (mccalluc) [inline rust/src/measurements/make_private_expr/expr_noise/mod.rs:187]:
```suggestion
                // smallest valid dataset distance is 2 in bounded-DP
                Some(MarginPub::Lengths) => 2,
```

File: rust/src/measurements/make_private_expr/expr_report_noisy_max/mod.rs [hunk 2]
```diff
@@ -71,7 +72,11 @@ where
         None => {
             // when scale is unknown, set it relative to the sensitivity of the query
             let margin = input_domain.active_margin().cloned().unwrap_or_default();
-            let (l_0, l_inf) = t_prior.map(&(margin.l_0(1), 1, margin.l_inf(1)))?;
+            let d_in = match margin.public_info {
+                Some(MarginPub::Lengths) => 2,
```
Comment (mccalluc) [inline rust/src/measurements/make_private_expr/expr_report_noisy_max/mod.rs:76]:
```suggestion
                // smallest valid dataset distance is 2 in bounded-DP
                Some(MarginPub::Lengths) => 2,
```

**Review by mccalluc — COMMENTED** [review]
Would it make sense to factor out the repeated code into a little helper function?

Also, we want the PRs to be as small as possible... but no smaller: Here I'm a little worried because I think the next PR in the stack
- #1834

is currently failing. For me, seeing this applied in the python code is a test that this is working as intended, so it's a bit of a worry if that test is failing. It would also be nice if it could be demonstrated without introducing an entire new notebook. I think `group-by.ipynb` is already in this branch: Could this change be exercised there?

Like we talked about on Thursday, there are different ways of making sure the Rust changes get followed up on in Python, and in different cases different approaches may be best.

File: rust/src/measurements/make_private_expr/expr_noise/test.rs [hunk 3]
```diff
@@ -134,3 +135,55 @@ fn test_make_laplace_grouped() -> Fallible<()> {
...
+            expr_domain,
+            PartitionDistance(InsertDeleteDistance),
+            MaxDivergence::default(),
+            Some(1.),
+        )?;
+
+    let epsilon = m_sum.map(&d_in)?;
+    // autocalibration chooses a noise scale based on the sensitivity.
+    // because of this, epsilon will always work out to 1.
+    println!("epsilon: {:?}", epsilon);
```
Comment (mccalluc) [inline rust/src/measurements/make_private_expr/expr_noise/test.rs:165]: Curious about this: Not doing any harm, except maybe for noise on stdout, but if there were a problem, wouldn't the `assert_eq!` tell us everything we needed?

**Review by mccalluc — APPROVED** [review]
Great!

File: rust/src/measurements/make_private_expr/expr_noise/test.rs (hunk 3 above)
Comment (Shoeboxam) [reply rust/src/measurements/make_private_expr/expr_noise/test.rs:165]: Yes, it's not necessary.

**Pull request merged**